concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs

```diff
@@ -385,17 +484,97 @@ markLiveBlockDead bp = do
+
+-- |Get the parent block of a live or finalized block. (For the genesis block, this will return
+-- the block itself.)
+parentOf ::
+    (LowLevel.MonadTreeStateStore m, MonadIO m, MonadState (SkovData (MPV m)) m) =>
+    BlockPointer (MPV m) ->
+    m (BlockPointer (MPV m))
+parentOf block
+    | Present blockData <- blockBakedData block = do
+        get >>= getBlockStatus (blockParent blockData) <&> \case
+            BlockAliveOrFinalized bp -> bp
+            _ ->
+                error $
+                    "parentOf: Parent block ("
+                        ++ show (blockParent blockData)
+                        ++ ") is not live or finalized."
+    | otherwise = return block
+
+-- |Get the parent block of a live (non-finalized) block.
+-- By definition, the parent block must either also be live or be the last finalized block.
+--
+-- If the block is not live, this function may fail with an error.
+parentOfLive :: HasCallStack => SkovData pv -> BlockPointer pv -> BlockPointer pv
+parentOfLive sd block
+    | let lastFin = sd ^. lastFinalized,
+      parentHash == getHash lastFin =
+        lastFin
+    | Just (MemBlockAlive parent) <- sd ^. blockTable . liveMap . at' parentHash = parent
+    | otherwise =
+        error $
+            "parentOfLive: parent block ("
+                ++ show parentHash
+                ++ ") is neither live nor last-finalized"
+  where
+    parentHash
+        | Present blockData <- blockBakedData block = blockParent blockData
+        | otherwise = error "parentOfLive: unexpected genesis block"
+
+-- |Determine if one block is an ancestor of another.
+-- A block is considered to be an ancestor of itself.
+isAncestorOf ::
+    (LowLevel.MonadTreeStateStore m, MonadIO m, MonadState (SkovData (MPV m)) m) =>
+    BlockPointer (MPV m) ->
+    BlockPointer (MPV m) ->
+    m Bool
+isAncestorOf b1 b2 = case compare (blockHeight b1) (blockHeight b2) of
+    GT -> return False
+    EQ -> return $ (getHash b1 :: BlockHash) == getHash b2
+    LT -> do
+        parent <- parentOf b2
+        b1 `isAncestorOf` parent
+
+-- * Operations on the branches
+
+-- |Add a newly-live block to the non-finalized branches.
+-- This assumes that the parent block is either the last finalized block or already among the
+-- non-finalized branches.
+-- The block should not already be present in the branches; if it is, a duplicate entry may be
+-- added.
+addToBranches :: (MonadState (SkovData pv) m) => BlockPointer pv -> m ()
+addToBranches block = do
+    lfbHeight <- use $ lastFinalized . to blockHeight
+    let insertIndex = fromIntegral $ blockHeight block - lfbHeight - 1
+    brs <- use branches
+    case compare insertIndex (Seq.length brs) of
+        LT -> branches . ix insertIndex %=! (block :)
+        EQ -> branches %=! (Seq.|> [block])
+        GT ->
+            error $
+                "Attempted to add a block at invalid height (" ++ show (blockHeight block) ++ ")"
```

Comment:
Shouldn't this precondition be mentioned?  Or is this implied by some other property stated somewhere else?

Reply:
Added a comment.